rust/ql/lib/codeql/rust/elements/internal/TraitImpl.qll

```diff
@@ -36,5 +36,36 @@ module Impl {
       not this.hasGenericParamList() and
       result = 0
     }
+
+    private int nrOfDirectTypeBounds() {
+      result = this.getTypeBoundList().getNumberOfBounds()
+      or
+      not this.hasTypeBoundList() and
+      result = 0
+    }
+
+    /**
+     * Gets the `index`th type bound of this trait, if any.
+     *
+     * This includes type bounds directly on the trait and bounds from any
+     * `where` clauses for `Self`.
+     */
+    TypeBound getTypeBound(int index) {
+      result = this.getTypeBoundList().getBound(index)
+      or
+      exists(WherePred wp |
+        wp = this.getWhereClause().getAPredicate() and
+        wp.getTypeRepr().(PathTypeRepr).getPath().getText() = "Self" and
+        result = wp.getTypeBoundList().getBound(index - this.nrOfDirectTypeBounds())
```

Comment:
It appears to be legal to have two where predicates with `Self` as the path, e.g.:
```
    trait Subtrait
    where
        Self: Supertrait1,
        Self: Supertrait2,
```
I think at the moment this will break your index numbering.

Reply:
You're right, that is legal and the implementation currently does not account for that.

Reply:
I've just created an issue tracking this.